### ISSUE TITLE
Indent with only one object

### ISSUE THREAD

**Issue by alallier** [eslint/eslint#3352]
eslint version `1.1.0`

Code example:

```
// Case 1
var object = { // Legal
  x: 2,
  y: 3
};

// Case 2  
var object2 = { // Not legal
      x: 2
    };

// Case 3   
var object3 = { // Legal
      x: 2
    },
    variable;
```

Rules:

```
"one-var": [2, "always"],
"indent": [2, 2, {"SwitchCase": 1, "VariableDeclarator": 2}]
```

eslint Output:

```
7:7  error  Expected indentation of 2 characters but found 6  indent
```

Case 2 should be legal
##

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by gyandeeps** [comment]
I think this was discussed and we agreed that case 2 will be illegal. There are so many variation people want for indentation is that its difficult to support all.

**Comment by kethinov** [comment]
I don't think there's a good reason that case 2 should be illegal while case 3 is legal.

Either both should be legal or neither should be legal.

Just adding another variable to the list should not completely change the alignment rules.

**Comment by nzakas** [comment]
@gyandeeps case 2 should be legal. If you're augmenting the indentation for `VariableDeclarator`, then that should apply even with object literals.

**Comment by BYK** [comment]
@nzakas - That makse case 1 illegal. Is that okay? This would break lots of existing code. (I tried, many tests fail)

**Comment by nzakas** [comment]
I think we need both to be legal.

**Comment by BYK** [comment]
Both being legal at the same time with the same options seems contradictory to me. That means "any indentation that is multiples of `x` is okay" which is not what this rule tries to enforce. At least as far as I understand.

**Comment by nzakas** [comment]
We have weird little exceptions like this all the time. Case 1 is the exception case, where if there's one object literal, it can be outdented. Case 2 is the normal case.

**Comment by BYK** [comment]
Alright, challenge accepted...

**Comment by alexindigo** [comment]
@BYK and continue our "comma-first" saga :)
this use case needs some love as well:

```
var bla     = 25
  , zap     = '123'
  , options = {
      method: 'get',
      port  : parsedUrl.port || 80,
      path  : parsedUrl.pathname,
      host  : parsedUrl.hostname
    }
  , boom    = 'test'
  ;
```

Thank you.

**Comment by alallier** [comment]
This still seems to be a problem is the latest version. I noticed also that this problem also occurs with arrays.

```
// Case 1
var object = [ // Legal
  x,
  y
];

// Case 2  
var object2 = [ // Not legal
      x
    ];

// Case 3   
var object3 = { // Legal
      x
    },
    variable;
```

**Comment by alexindigo** [comment]
To add to the list of broken things:

```
var a =
{
  bb: 1
};
```

Doesn't work either.

**Comment by JamesMGreene** [comment]
I'm hitting this as well with the latest version, with the initial comment on this issue still holding true. :cry:

**Comment by sstern6** [comment]
@nzakas this seems related to an issue im trying to patch right now https://github.com/eslint/eslint/issues/4161.  Waiting for feedback to proceed on the issue.  Id be happy to take a swing at this one as well if the community would like. If you'd like them separate I can tackle this once the other PR is up and ready!

Thanks and let me know how you reccomend proceeding.

**Comment by nzakas** [comment]
Sure, I think a PR could help move this along.

**Comment by sstern6** [comment]
@nzakas based on the conversation going on here... https://github.com/eslint/eslint/issues/7274. I will wait until the indent rewrite is complete before continuing on indent related issues.  If you have any thoughts on that, I would appreciate the feedback.

**Comment by kaicataldo** [comment]
Is this addressed by the `indent` rule rewrite?

**Comment by not-an-aardvark** [comment]
No, the `indent` rewrite behaves the same as the current rule for these cases.

For what it's worth, I think this is working as intended.

**Comment by kethinov** [comment]
It's an accepted bug though, so I wouldn't say it's working as intended.

**Comment by not-an-aardvark** [comment]
We're currently trying to minimize the number of cases where the `indent`  rule's behavior is "any indentation is ok here", particularly in the rewritten version of the rule. While it's clear that some people would prefer the behavior recommended in this issue, there are many other people (myself included) who prefer the indentation that the rule currently enforces. Therefore, we have three feasible options:

* Keep the behavior as-is.
* Allow multiple indentation levels in cases like this. While this would be doable (by simply ignoring these lines), I don't think it's a good idea for the rule to ignore lines unless there's a compelling reason to do so. I would be particularly wary of ignoring lines in this case because there is no backwards-compatibility benefit from doing so.
* Add an option for this. This is also doable, but the `indent` rule already has a fairly large number of options, and I'm not sure this is an important enough issue to merit making the schema even more confusing to new users.

**Comment by andyearnshaw** [comment]
@not-an-aardvark coding style diversity isn't important enough?

FWIW, I choose this code style because it's more readable in certain situations.  The hanging indent makes a clear distinction on the current line _at the beginning_, rather than at the end or on preceding or following lines.

ESLint forces inconsistency over single and multiple assignments like this (with `indent: 4`):

```js
// Invalid
let outerScopeVar,
    aReallyLongNamedFunctionCall = () => next(),
    something = doSomething().then(doSomethingElse).then(() => {
    outerScopeVar = true;
    aReallyLongNamedFunctionCall()
});

// Must do this
let outerScopeVar,
    aReallyLongNamedFunctionCall = () => next(),
    something = doSomething().then(doSomethingElse).then(() => {
        outerScopeVar = true;
        aReallyLongNamedFunctionCall()
    });
```

```js
// Invalid
let singleAssignment = doSomething().then(doSomethingElse).then(() => {
        outerScopeVariable = true;
    });

// Must do this
let singleAssignment = doSomething().then(doSomethingElse).then(() => {
    outerScopeVariable = true;
});
```

I'm sure you can agree, the "invalid" code in the first block looks hideous, so it seems fair to enforce the hanging indent.  But if I want to be consistent and use a hanging indent everywhere, I'm not allowed.

**Comment by ilyavolodin** [comment]
@andyearnshaw Unfortunately, indentation is a very personal thing. For example, you are thinking that first example you provided is hideous, and I think the way you propose to write it is unreadable and I would hate to see that style in my codebase.
Problem is, `indent` is the most complex rule we have. We already re-written it 3 times completely from the scratch, and, while we would love to support every style, we just can't:-( So we have to significantly limit number of options and styles we support.

**Comment by andyearnshaw** [comment]
@ilyavolodin I fully agree it is a personal thing and I also agree that you can't accommodate everyone.  My point is that, as things stand now, ESLint _forces you_ to be inconsistent if you declare multiple variables per statement.

Though I disagree with you on which version is "unreadable", the rewrite for the first example that I gave wasn't a proposal.  It is the way you are forced to write it under the current rules unless you put each variable declaration in its own statement.

**Comment by kaicataldo** [comment]
Closing this enhancement proposal since it is working as intended and there hasn't been any discussion in quite a while. While we wish we could accommodate all requests, we have limited resources and do need to prioritize.

**Comment by andyearnshaw** [comment]
@kaicataldo that's disappointing, I really don't think that it should be considered "working as intended".  That being said, I understand that this bug isn't a priority for the eslint team.  If I get time in the future, would you be willing to accept a pull request?

**Comment by kaicataldo** [comment]
Our policy is that issues require 3 👍s and a champion (someone who's willing to see the proposal through to completion) from the team. This can happen in either an issue or a pull request. 

We generally recommend making an issue and waiting to write code until it's accepted (so as to avoid taking the time to write code that won't be used) but having a working implementation can make it easier to discuss/advocate for because the work is done already.

All that to say that you can feel free to make a new issue (this would be an enhancement proposal for a new option rather than a bug fix) or PR and we can see what the team thinks.

**Comment by JamesMGreene** [comment]
Meh, very disheartening. :confused:

**Comment by platinumazure** [comment]
We're perfectly willing to accept new, concrete proposals for how to fix this (assuming we get support from the team). If someone wants to create a new issue with a concrete proposal for how best to fix this-- taking into account people's *wildly* differing preferences on how to handle indentation of object/array expressions in single or multiple variable declarators-- then we are happy to evaluate.

[kaicataldo's comment above](https://github.com/eslint/eslint/issues/3352#issuecomment-312273272) explains our policy at a high level. We don't aim to stifle discussion or (unreasonably) reject ideas. We just need to maintain our issue backlog and try to close issues that aren't going anywhere to keep ourselves sane. But we are happy to look at new issues, especially ones that have a new take on an older problem.

**Comment by kaicataldo** [comment]
Yes, new issues and PRs are always welcome! Apologies if that didn't come across in my response. We are a small team of volunteers and we have policies in place in an effort to keep our backlog of issues maintainable.
